include/sys/scheduler.h

```diff
@@ -0,0 +1,171 @@
+ * management of IPC objects, either in the core kernel or other IPC
+ * implemented by OS compatibility layers, providing basic wait/wake operations
+ * with spinlocks used for synchronization.
+ *
+ * These APIs are public and will be treated as contract, even if the
+ * underlying scheduler implementation changes.
+ */
+
+/**
+ * Callback function prototype for k_sched_wake_one/wake_all
+ *
+ * If a thread is woken up, the callback function is called with the thread
+ * as an argument. The sched_spinlock is held the entire time, ensuring that
+ * thread state doesn't suddenly change (from some nested IRQ or activity on
+ * another CPU) when the callback is modifying the thread's state.
+ *
+ * It is only inside these callbacks that it is safe to inspect or modify
+ * the thread that was woken up.
+ */
+typedef void (*k_sched_wake_cb_t)(struct k_thread *thread, void *obj,
+				  void *context);
+
+/**
+ * Wake up a thread pending on the provided wait queue, with callback
+ *
+ * Given a wait_q, wake up the highest priority thread on the queue. If the
+ * queue was empty just return false.
+ *
+ * Otherwise, do the following, in order,  holding sched_spinlock the entire
+ * time so that the thread state is guaranteed not to change:
+ * - if the callback function is provided, run the callback function with
+ *   the provided object and context pointers
+ * - Set the thread's swap return values to swap_retval and swap_data
+ * - un-pend and ready the thread, but do not invoke the scheduler.
+ *
+ * It is up to the caller to implement locking such that the return value of
+ * this function (whether a thread was woken up or not) does not immediately
+ * become stale.
+ *
+ * Repeated calls to this function until it returns false is a suitable
+ * way to wake all threads on the queue.
+ *
+ * It is up to the caller to implement locking such that the return value of
+ * this function does not immediately become stale. Calls to wait and wake on
+ * the same wait_q object must have synchronization. Calling this without
+ * holding any spinlock is a sign that this API is not being used properly.
+ *
+ * @param wait_q Wait queue to wake up the highest prio thread
+ * @param swap_retval Swap return value for woken thread
+ * @param swap_data Data return value to supplement swap_retval. May be NULL.
+ * @param cb Callback function, or NULL in which case no callback is invoked.
+ * @param obj Kernel object associated with this wake operation, passed to cb.
+ *        Only used by the callback. May be NULL.
+ * @param context Additional context pointer associated with this wake
+ *        operation. Only used by the callback. May be NULL.
+ * @retval true If a thread waa woken up
+ * @retval false If the wait_q was empty
+ */
+bool k_sched_callback_wake(_wait_q_t *wait_q, int swap_retval, void *swap_data,
+			   k_sched_wake_cb_t cb, void *obj, void *context);
+
+/**
+ * Wake up a thread pending on the provided wait queue
```

Comment:
Should document as "wake up the highest priority thread" I think.  Strictly a wait queue will present the highest priority thread that has been in the queue the longest for wakeup at the head of the list.  We never had to detail that expressly anywhere because it was an internal API, but it deserves a callout here I think.

Reply:
Yes agreed, will add this.